decimal64.go

```diff
@@ -102,34 +174,63 @@ func renormalize(exp int, significand uint64) (int, uint64) {
 }
 
 func rescale(exp int, significand uint64, targetExp int) (uint64, int) {
-	exp -= targetExp
-	var divisor uint64 = 1
-	for ; exp < -7 && divisor < significand; exp += 8 {
-		divisor *= 100000000
-	}
-	for ; exp < -3 && divisor < significand; exp += 4 {
-		divisor *= 10000
-	}
-	for ; exp < -1 && divisor < significand; exp += 2 {
-		divisor *= 100
-	}
-	for ; exp < 0 && divisor < significand; exp++ {
-		divisor *= 10
+	expDiff := targetExp - exp
+	mag := size(significand)
+	if targetExp-exp > mag {
+		return 0, targetExp
 	}
+	divisor := powersOf10[expDiff]
 	return significand / divisor, targetExp
 }
 
-func matchScales(exp1 int, significand1 uint64, exp2 int, significand2 uint64) (int, uint64, int, uint64) {
-	if significand1 == 0 {
-		exp1 = exp2
-	} else if significand2 == 0 {
-		exp2 = exp1
-	} else if exp1 < exp2 {
-		significand1, exp1 = rescale(exp1, significand1, exp2)
-	} else if exp2 < exp1 {
-		significand2, exp2 = rescale(exp2, significand2, exp1)
+func (dec *decParts) rescale(targetExp int) {
+	expDiff := targetExp - dec.exp
+	mag := dec.mag
+	if expDiff > mag {
+		dec.significand, dec.exp = 0, targetExp
+		return
+	}
+	divisor := powersOf10[expDiff]
+	dec.significand = dec.significand / divisor
+	dec.exp = targetExp
+}
+
+// roundHalfUp implements the half up rounding attribute to a significand which can't be fully stored in 16 decimal digits.
+func roundHalfUp(significand uint64, exp int, targetExp int) bool {
+	expDiff := targetExp - exp
+	divisor := powersOf10[expDiff]
+	resizedSig := significand / divisor
+	truncatedSig := resizedSig * divisor
+	return significand-truncatedSig >= 5*powersOf10[expDiff-1]
+}
+
+func roundDefault(significand uint64, exp int, targetExp int) bool {
+	return true
+}
+
+// roundHalfUp implements the half up rounding attribute to a significand which can't be fully stored in 16 decimal digits.
+func roundZeroUp(significand uint64, exp int, targetExp int) bool {
+	expDiff := targetExp - exp
+	divisor := powersOf10[expDiff]
+	resizedSig := significand / divisor
+	truncatedSig := resizedSig * divisor
+	return significand-truncatedSig >= powersOf10[expDiff-1]
+}
+
+// match scales matches the exponents of d and e and uses dec to get the rounding status
+// decs rounding method must be set or panic will occur
+func (dec *decParts) matchScales(d, e *decParts) {
```

Comment:
`dec` is being used merely for its rounding context, and it seems that this is the only use of the relevant fields in `decParts`. I think you can just pass that context in and remove rounding fields from `decParts`.